metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteEntityService.java

```diff
@@ -0,0 +1,254 @@
+
+    final List<RelatedAspect> relatedAspects = relatedEntities.getEntities().stream().map(relatedEntity -> {
+          final Urn relatedUrn = UrnUtils.getUrn(relatedEntity.getUrn());
+          final String relationType = relatedEntity.getRelationshipType();
+          final String relatedEntityName = relatedUrn.getEntityType();
+          final EntitySpec relatedEntitySpec = _entityService.getEntityRegistry().getEntitySpec(relatedEntityName);
+
+          return getAspectReferringTo(urn, relatedUrn, relationType, relatedEntitySpec).map(pair -> {
+            final RelatedAspect relatedAspect = new RelatedAspect();
+            relatedAspect.setEntity(relatedUrn);
+            relatedAspect.setRelationship(relationType);
+            relatedAspect.setAspect(pair.getSecond().getName());
+            return relatedAspect;
+          });
+        }).filter(java.util.Optional::isPresent)
+        .map(java.util.Optional::get)
+        .limit(10).collect(Collectors.toList());
+
+    result.setRelatedAspects(new RelatedAspectArray(relatedAspects));
+    result.setTotal(relatedEntities.getTotal());
+
+    if (dryRun) {
+      return result;
+    }
+
+    int processedEntities = 0;
+
+    // Delete first 10k
+    relatedEntities.getEntities().forEach(entity -> deleteReference(urn, entity));
+
+    processedEntities += relatedEntities.getEntities().size();
+
+    // Delete until less than 10k are left
+    while (relatedEntities.getCount() > 10000) {
+      log.info("Processing batch {} of {} aspects", processedEntities, relatedEntities.getTotal());
+      sleep(ELASTIC_BATCH_DELETE_SLEEP_SEC);
+      relatedEntities = _graphService.findRelatedEntities(null, newFilter("urn", urn.toString()),
+          null, EMPTY_FILTER, ImmutableList.of(),
+          newRelationshipFilter(EMPTY_FILTER, RelationshipDirection.INCOMING), 0, 10000);
+      relatedEntities.getEntities().forEach(entity -> deleteReference(urn, entity));
+      processedEntities += relatedEntities.getEntities().size();
+    }
+
+    return result;
+  }
+
+  /**
+   * Utility method to sleep the thread.
+   * @param seconds The number of seconds to sleep.
+   */
+  private void sleep(final Integer seconds) {
+    try {
+      TimeUnit.SECONDS.sleep(seconds);
+    } catch (InterruptedException e) {
+      log.error("Interrupted sleep", e);
+    }
+  }
+
+  /**
+   * Processes an aspect of a given {@link RelatedEntity} instance that references a given {@link Urn}, removes said
+   * urn from the aspects and submits an MCP with the updated aspects.
+   * @param urn           The urn to be found.
+   * @param relatedEntity The entity to be modified.
+   */
+  private void deleteReference(final Urn urn, final RelatedEntity relatedEntity) {
+    final Urn relatedUrn = UrnUtils.getUrn(relatedEntity.getUrn());
+    final String relationType = relatedEntity.getRelationshipType();
+    final String relatedEntityName = relatedUrn.getEntityType();
```

Comment:
Why do we need to redefine these variables here? Not clear to me - just readability?

Reply:
Just readability